File: GravityBomb.java
public class GravityBomb {
    String[] aftermath(String[] board) {
        int w = board[0].length(), 
            h = board.length;
        
        int[] depths = new int[w]; int min = 51;
        for (int x = 0; x < w; ++x) {
            depths[x] = 0;
            for (int y = 0; y < h; ++y) 
                if (board[y].charAt(x) == 'X') ++depths[x];
            if (depths[x] < min) min = depths[x];
        }
        for (int i = 0; i < w; ++i)
            depths[i] = h - (depths[i] - min);
        
        char[][] temp = new char[h][w];
        for (int x = 0; x < w; ++x) {
            for (int y = 0; y < depths[x]; ++y)
                temp[y][x] = '.';
            for (int y = depths[x]; y < h; ++y)
                temp[y][x] = 'X';
        }

        String[] ret = new String[h];
        for (int i = 0; i < h; ++i)
            ret[i] = new String(temp[i]);
        return ret;
    }
}

File: D2L2/GravityBomb.cpp
#include <vector>
#include <string>
using namespace std;

vector<string> aftermath(vector<string> const &board) {
    int w = board[0].size(), 
        h = board.size();

    vector<int> depths; int dmin = 51;
    for (int x = 0; x < w; ++x) {
        int depth = 0;
        for (int y = 0; y < h; ++y)
            if (board[y][x] == 'X') ++depth;            
        depths.push_back(depth);
        if (depth < dmin) dmin = depth;
    }
    for (int i = 0; i < depths.size(); ++i)
        depths[i] -= dmin;

    string none(w, '.');
    vector<string> ret(h, none);
    for (int x = 0; x < w; ++x)
        for (int y = 0; y < depths[x]; ++y)
            ret[h - y - 1][x] = 'X';
    return ret;
}